crates/hir-def/src/data/adt.rs

```diff
@@ -317,6 +318,23 @@ impl EnumData {
             _ => IntegerType::Pointer(true),
         }
     }
+
+    pub fn is_payload_free(&self, db: &dyn DefDatabase) -> bool {
+        self.variants.iter().all(|(v, _)| {
+            let variant = &db.enum_variant_data(*v).variant_data;
+            if !variant.fields().is_empty() {
+                return false;
+            }
+            if !matches!(variant.kind(), StructKind::Unit) {
+                let body = db.body((*v).into());
+                // A variant with explicit discriminant
+                if body.exprs[body.body_expr] != Expr::Missing {
```

Comment:
shouldn't the kind check suffice? Why query the body

Reply:
Because the condition is
Not having a const constructor && has a explicit discriminant

https://github.com/rust-lang/rust/blob/bd53aa3bf7a24a70d763182303bd75e5fc51a9af/compiler/rustc_middle/src/ty/adt.rs#L446-L448

The kind check is for the former and the body query is for explicit discriminant, though the order is opposite to rustc to mitigate perf issue

Reply:
Ah okay, let's copy the comments over then

Reply:
Maybe we could apply the similar heuristic if the body query causes noticeable perf regression like in my comment below and that would make some false negatives, not false positives errors.
Or it might be possible to specialized function for quickly checking whether the enum variant has explicit discriminant or not